File: day1/two_sum.go
package main

// twoSum finds two numbers in nums that add up to target and returns their indices.
func twoSum(nums []int, target int) []int {
	// Create a map to store the complement of each number and its index.
	numMap := make(map[int]int)

	// Iterate through the array.
	for i, num := range nums {
		// Calculate the complement needed to reach the target.
		complement := target - num

		// Check if the complement exists in the map.
		if j, found := numMap[complement]; found {
			// Return the indices of the two numbers.
			return []int{j, i}
		}

		// Store the current number and its index in the map.
		numMap[num] = i
	}

	// If no solution is found, return an empty slice (though by problem statement there will always be a solution).
	return []int{}
}

File: prakash/03_Twosum/twosum.ts
function twosum(nums: number[], target: number): number[] | null {
  // Initialize an empty map to store numbers and their indices
  const numMap: { [key: number]: number } = {};

  // Iterate over the array of numbers

  for (let i = 0; i < nums.length; i++) {
    // Calculate the complement of the current number

    const complement: number = target - nums[i];

    // Check if the complement exists in the map

    if (numMap.hasOwnProperty(complement)) {
      // hasOwnProperty return boolean has its checks if the complement already exists in numMap;

      // If it exists, return the indices of the two numbers

      return [numMap[complement], i];
    }

    // Store the current number and its index in the map

    numMap[nums[i]] = i;
  }

  // If no solution is found, throw Error
  throw new Error("No two sum solution");
}
